core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala

```diff
@@ -193,17 +193,21 @@ class HadoopRDD[K, V](
   override def getPartitions: Array[Partition] = {
     val jobConf = getJobConf()
     // add the credentials here as this can be called before SparkContext initialized
+    // Also need to explicitly run this as the Spark user to pick up permissions -
+    // the driver does not invoke runAsSparkUser elsewhere, unlike the executors
     SparkHadoopUtil.get.addCredentials(jobConf)
-    val inputFormat = getInputFormat(jobConf)
-    if (inputFormat.isInstanceOf[Configurable]) {
-      inputFormat.asInstanceOf[Configurable].setConf(jobConf)
-    }
-    val inputSplits = inputFormat.getSplits(jobConf, minPartitions)
-    val array = new Array[Partition](inputSplits.size)
-    for (i <- 0 until inputSplits.size) {
-      array(i) = new HadoopPartition(id, i, inputSplits(i))
-    }
-    array
+    SparkHadoopUtil.get.runAsSparkUser(() => {
```

Comment:
Ah - if the YARN Client executes this code, it already inherits the permissions set up by the YARN environment, which makes running this unnecessary. Whereas if this is run in standalone mode, we need to explicitly run this code as the YARN User at the driver.

I don't want to wrap all of SparkContext.runJob in a runAsSparkUser call. We might as well localize the things that need HDFS permissions to the bits of code that actually need it.